### ISSUE TITLE
Question - what would make `LintDriver` take an extremely long time to run?

### ISSUE THREAD

**Issue by saiimons** [google/android-lint-performance-probe#5]
After updating a project from gradle `5.4.1` to `5.6.4` and `com.android.tools.build:gradle` from `3.5.2` to `3.6.2`, running lint went from a couple minutes to 20+

Lint block looks like this:
```
{
    abortOnError false
    htmlReport true
    textReport false
    xmlReport true
}
```

I ran the profiler and this is the output I now have:
```
Lint detector performance stats:
                                                    total           self          calls
                                LintDriver     1535684 ms     1468709 ms          77536
                            GradleDetector       42737 ms       42737 ms           1664
               TopDownAnalyzerFacadeForJVM       22158 ms       22158 ms              7
                          OverdrawDetector         677 ms         677 ms           1836
                              IconDetector         277 ms         277 ms           6508
                         CallSuperDetector         111 ms         111 ms          15504
                    InvalidPackageDetector         107 ms         107 ms         468259
                           TypedefDetector          86 ms          86 ms           1515
               AppCompatCustomViewDetector          78 ms          78 ms             61
                      ResourceTypeDetector          69 ms          69 ms           8336
                     ByteOrderMarkDetector          65 ms          65 ms           1298
                            ThreadDetector          61 ms          61 ms           1143
                               ApiDetector          54 ms          54 ms          93218
                           CleanupDetector          46 ms          46 ms            457
                   PrivateResourceDetector          37 ms          37 ms          25702
                        PrivateKeyDetector          37 ms          37 ms          11358
                    UnusedResourceDetector          37 ms          36 ms          59746
                      StringFormatDetector          33 ms          33 ms           4912
                       TranslationDetector          30 ms          29 ms          14590
                      MissingClassDetector          29 ms          26 ms          18183
                              TypoDetector          22 ms          22 ms           8844
                     ResourceCycleDetector          13 ms          13 ms          10846
                       MergeMarkerDetector          13 ms          13 ms           5683
                        PrivateApiDetector          12 ms          12 ms             37
                          ViewTypeDetector          12 ms          12 ms           1454

#######################################
##### A lot of really small calls #####
#######################################

                                         -              -              -              -
                                     total     1535745 ms     1535745 ms         983166
```

Any clue on how to diagnosis the issue ?

EDIT: FYI, report contains 9 errors & 427 warnings
EDIT2: no fancy configuration, ran lint on a single module, for a single build configuration.

**Comment by saiimons** [comment]
With the previous plugin versions:
```
Lint detector performance stats:
                                                    total           self          calls
                                LintDriver       55472 ms       29142 ms          83560
               TopDownAnalyzerFacadeForJVM       23571 ms       23571 ms              7
                          OverdrawDetector        1040 ms        1040 ms           1835
                      ResourceTypeDetector         200 ms         200 ms           6189
                    AppIndexingApiDetector         171 ms         171 ms            187
                              IconDetector         157 ms         156 ms           6507
                    InvalidPackageDetector         124 ms         124 ms         465132
                         CallSuperDetector         117 ms         117 ms          15477
                           TypedefDetector         103 ms         103 ms           1471
                               ApiDetector          75 ms          75 ms          90359
                     ByteOrderMarkDetector          68 ms          68 ms           1297
                           CleanupDetector          60 ms          60 ms            442
               AppCompatCustomViewDetector          56 ms          56 ms            163
                        PrivateKeyDetector          49 ms          49 ms          13935
                            GradleDetector          42 ms          41 ms           1483
                    UnusedResourceDetector          39 ms          39 ms          59674
                   PrivateResourceDetector          39 ms          39 ms          25725
                      StringFormatDetector          34 ms          34 ms           4326
                       TranslationDetector          31 ms          30 ms          14590
                      MissingClassDetector          30 ms          27 ms          18142
                            ThreadDetector          29 ms          29 ms           1897
                              TypoDetector          26 ms          26 ms           8844
                     ResourceCycleDetector          17 ms          17 ms          10846
                          ViewTypeDetector          15 ms          15 ms           1440
                        PrivateApiDetector          14 ms          14 ms             37
                       MergeMarkerDetector          13 ms          13 ms           6973
                        PermissionDetector          12 ms          12 ms            127

#######################################
##### A lot of really small calls #####
#######################################

                                         -              -              -              -
                                     total       55496 ms       55496 ms         959806

```

**Comment by saiimons** [comment]
Memory comparison

Before:
```
Lint detector performance stats:
                                                    total           self          calls
                                LintDriver       10105 MB        5841 MB          83560
               TopDownAnalyzerFacadeForJVM        3429 MB        3429 MB              7
                          OverdrawDetector         485 MB         485 MB           1835
                      ResourceTypeDetector          92 MB          92 MB           6189
               AppCompatCustomViewDetector          38 MB          38 MB            163
                           TypedefDetector          36 MB          36 MB           1471
                         CallSuperDetector          31 MB          31 MB          15477
                           CleanupDetector          23 MB          23 MB            442
                              IconDetector          15 MB          15 MB           6507
                      StringFormatDetector          15 MB          15 MB           4326
                               ApiDetector          13 MB          13 MB          90359
                    InvalidPackageDetector          12 MB          12 MB         465132
                        PrivateApiDetector          12 MB          12 MB             37
                     ByteOrderMarkDetector           8 MB           8 MB           1297
                   PrivateResourceDetector           7 MB           7 MB          25725
                      MissingClassDetector           5 MB           5 MB          18142
                        PrivateKeyDetector           5 MB           5 MB          13935
                            GradleDetector           4 MB           4 MB           1483
                       TranslationDetector           3 MB           3 MB          14590
                               RtlDetector           2 MB           2 MB           3670
                              LeakDetector           2 MB           2 MB            466
                    AppIndexingApiDetector           2 MB           2 MB            187
                    UnusedResourceDetector           2 MB           2 MB          59674
                       MergeMarkerDetector           2 MB           2 MB           6973
                  ConstraintLayoutDetector           1 MB           1 MB             30
                            ThreadDetector           1 MB           1 MB           1897
                     ResourceCycleDetector           1 MB           1 MB          10846

#######################################
#####         A lot of 0MB        #####
#######################################

                                         -              -              -              -
                                     total       10110 MB       10110 MB         959806
```

After:
```
Lint detector performance stats:
                                                    total           self          calls
                                LintDriver      300979 MB      297202 MB          77532
               TopDownAnalyzerFacadeForJVM        3344 MB        3344 MB              7
                          OverdrawDetector         132 MB         132 MB           1836
               AppCompatCustomViewDetector          89 MB          89 MB             61
                           TypedefDetector          32 MB          32 MB           1507
                         CallSuperDetector          32 MB          32 MB          15470
                              IconDetector          15 MB          15 MB           6508
                           CleanupDetector          15 MB          15 MB            457
                      StringFormatDetector          14 MB          14 MB           4912
                               ApiDetector          13 MB          13 MB          93205
                    InvalidPackageDetector          12 MB          12 MB         468259
                        PrivateApiDetector          12 MB          12 MB             37
                            ThreadDetector           8 MB           8 MB            940
                     ByteOrderMarkDetector           8 MB           8 MB           1298
                   PrivateResourceDetector           8 MB           8 MB          25702
                      ResourceTypeDetector           5 MB           5 MB           8336
                      MissingClassDetector           5 MB           5 MB          18183
                        PrivateKeyDetector           4 MB           4 MB          11358
                            GradleDetector           4 MB           4 MB           1664
                       TranslationDetector           3 MB           3 MB          14590
                               RtlDetector           2 MB           2 MB           3676
                              LeakDetector           2 MB           2 MB            467
                    UnusedResourceDetector           2 MB           2 MB          59746
                       MergeMarkerDetector           2 MB           2 MB           5683
                  ConstraintLayoutDetector           1 MB           1 MB             30
                     ResourceCycleDetector           1 MB           1 MB          10846

#######################################
#####         A lot of 0MB        #####
#######################################

                                         -              -              -              -
                                     total      300992 MB      300992 MB         982907

```

**Comment by gharrma** [comment]
Those are some interesting numbers!

LintDriver is basically the thing that prepares the environment, parses source files, and dispatches to the detectors. So it's hard to diagnose the problem from this data alone.

If you like, you can capture a YourKit CPU sampling snapshot and send it to me for analysis. To do so, you can add this to your Gradle JVM args:

```
-agentpath:/path/to/yourkit/bin/linux-x86-64/libyjpagent.so=disablealloc,disabletracing,probe_disable=*,exceptions=disable,sampling,onexit=snapshot
```

Make sure to change the agentpath to match your YourKit installation and platform---you can reuse the path that you used for the probe tool. Also, use Gradle's `--no-daemon` flag so that the JVM args take effect, and so that the JVM shuts down afterward and the snapshot gets written.

Once the Gradle daemon exits, the snapshot should show up under `~/Snapshots`. Let me know if you run into any issues. You can share the snapshot here or send it to gharrma@google.com

**Comment by saiimons** [comment]
@gharrma thanks for the quick reply, I'll run this and brew myself a coffee ☕️ , I'll send you the results once they're available.

**Comment by saiimons** [comment]
@gharrma just send you the data!

**Comment by gharrma** [comment]
Thanks! Copying some points from my email, for others:
- The regression is likely caused by [IDEA-229655](https://youtrack.jetbrains.com/issue/IDEA-229655)
- It is fixed in AGP 4.0. See [b/144513656#comment13](https://issuetracker.google.com/144513656#comment13) for an explanation of why the fix is not backported to 3.6, along with possible workarounds.

**Comment by giangpham96** [comment]
Hi there, I'm using AGP 4.1.3, seems like `LintDriver` & `TopDownAnalyzerFacadeForJVM` still takes time to run. Would you consider this slow? @gharrma ?
```
Lint detector performance stats:
                                                    total           self          calls
                                LintDriver      254527 ms       66103 ms         850893
               TopDownAnalyzerFacadeForJVM      180808 ms      180808 ms           2412
                     ByteOrderMarkDetector        3085 ms        3084 ms           8266
                      MissingClassDetector        1129 ms        1129 ms          80970
                              IconDetector         537 ms         537 ms           9235
                   PrivateResourceDetector         513 ms         429 ms         388517
                    UnusedResourceDetector         384 ms         376 ms          16943
                       MergeMarkerDetector         291 ms         291 ms          15300
                               ApiDetector         185 ms         181 ms         384301
                       TranslationDetector         175 ms         167 ms         289308
                      StringFormatDetector         165 ms         165 ms         199349
                         CallSuperDetector          91 ms          91 ms          35152
                      ResourceTypeDetector          86 ms          86 ms           4692
                             RangeDetector          83 ms          83 ms           1020
                 DuplicateResourceDetector          81 ms          81 ms         134971
                        PrivateKeyDetector          78 ms          78 ms          24771
                          OverdrawDetector          70 ms          69 ms          13189
                           TypedefDetector          58 ms          58 ms            870
                     ResourceCycleDetector          52 ms          51 ms         197550
                      LocaleFolderDetector          44 ms          44 ms         161153
                            ButtonDetector          42 ms          36 ms         105954
                         NamespaceDetector          33 ms          33 ms          78814
                        TypographyDetector          32 ms          32 ms         104059
                               RtlDetector          30 ms          27 ms          19973
                        PermissionDetector          29 ms          27 ms            398
                            ThreadDetector          26 ms          26 ms            380
                        VectorPathDetector          25 ms          22 ms           2705
                            VectorDetector          23 ms          20 ms           8028
                           PxUsageDetector          20 ms          20 ms         130689
                        CanvasSizeDetector          17 ms          17 ms            310
                     AccessibilityDetector          17 ms          11 ms            561
        ClickableViewAccessibilityDetector          16 ms          15 ms            705
                          ManifestDetector          16 ms          14 ms           1587
                           WrongIdDetector          15 ms          15 ms          18611
                  ConstraintLayoutDetector          15 ms          15 ms            310
                           CleanupDetector          13 ms          13 ms            556
                    CordovaVersionDetector          13 ms          13 ms         166302
                  UnsafeNativeCodeDetector          12 ms          12 ms            650
                            ParcelDetector          12 ms          12 ms            429
                   LayoutInflationDetector          11 ms          11 ms           1112
                  ManifestResourceDetector          11 ms          11 ms           6615
                       DuplicateIdDetector          11 ms          10 ms           8223
          TrustAllX509TrustManagerDetector          10 ms          10 ms         157661
               IgnoreWithoutReasonDetector           9 ms           9 ms           1332
                       CheckResultDetector           9 ms           7 ms           2954
                           SetTextDetector           7 ms           7 ms           1882
                         ExtraTextDetector           7 ms           7 ms          26863
              ObsoleteLayoutParamsDetector           7 ms           7 ms          45027
                 LayoutConsistencyDetector           6 ms           6 ms          11037
                          ViewTypeDetector           6 ms           6 ms           3730
                              Utf8Detector           6 ms           6 ms           3402
            WebViewApiAvailabilityDetector           5 ms           5 ms          29660
                     ScopedStorageDetector           5 ms           3 ms            260
                            IntentDetector           5 ms           5 ms            269
                           PluralsDetector           5 ms           4 ms            343
                            GradleDetector           5 ms           5 ms           1720
                 InefficientWeightDetector           4 ms           4 ms            961
                       DataBindingDetector           4 ms           4 ms          62119
                            LocaleDetector           4 ms           3 ms            299
                              LeakDetector           4 ms           4 ms           1439
                          TextViewDetector           4 ms           4 ms           1122
                   JavaPerformanceDetector           4 ms           4 ms           1284
                   ViewConstructorDetector           4 ms           4 ms            376
               AppCompatCustomViewDetector           4 ms           4 ms            310
                          CutPasteDetector           4 ms           4 ms           1681
                          ChromeOsDetector           4 ms           4 ms            764
                KeyboardNavigationDetector           4 ms           3 ms            132
                       UselessViewDetector           4 ms           3 ms           1826
                   RelativeOverlapDetector           3 ms           3 ms             99
                       DeprecationDetector           3 ms           3 ms           1637
                    ResourcePrefixDetector           3 ms           3 ms           9432
                AbstractAnnotationDetector           3 ms           0 ms             49
                            AssertDetector           3 ms           3 ms           2306
                    ObjectAnimatorDetector           2 ms           2 ms            480
                   HardcodedValuesDetector           2 ms           2 ms           1012
                 CipherGetInstanceDetector           2 ms           2 ms            268
                     AppLinksValidDetector           2 ms           2 ms            495
                        AnnotationDetector           2 ms           2 ms           1315
                        HardwareIdDetector           2 ms           2 ms           1534
                     ExifInterfaceDetector           2 ms           2 ms           1284
                          SecurityDetector           2 ms           2 ms           1867
              MergeRootFrameLayoutDetector           2 ms           2 ms           1524
                       WrongImportDetector           2 ms           2 ms           1245
             NonConstantResourceIdDetector           2 ms           2 ms           1269
               JavaScriptInterfaceDetector           1 ms           1 ms           1528
             NetworkSecurityConfigDetector           1 ms           1 ms            185
                          IteratorDetector           1 ms           1 ms            426
                   TranslucentViewDetector           1 ms           1 ms           2911
                         StateListDetector           1 ms           1 ms            847
                      ManifestTypoDetector           1 ms           1 ms            942
                         AndroidTvDetector           1 ms           1 ms            881
                          AutofillDetector           1 ms           1 ms             68
                        InstantAppDetector           1 ms           1 ms           1274
               UseCompoundDrawableDetector           1 ms           1 ms            500
                    StringAuthLeakDetector           1 ms           1 ms           1165
                            SdCardDetector           1 ms           1 ms           1165
                              FontDetector           1 ms           1 ms            159
                      TooManyViewsDetector           1 ms           1 ms           9350
                        DateFormatDetector           1 ms           1 ms            496
                          LabelForDetector           1 ms           1 ms            913
                 PackageVisibilityDetector           1 ms           1 ms            503
                  AlwaysShowActionDetector           1 ms           1 ms           2488
...
```

**Comment by gharrma** [comment]
Hi @giangpham96. Most likely it is a separate issue from the one discussed earlier (IDEA-229655). `TopDownAnalyzerFacadeForJVM` is basically the Kotlin compiler frontend analyzing your code, and it is not unusual for this to take a long time if you have a large projects with lots of Kotlin code (especially if you use `checkDependencies = true`). There are some improvements coming in AGP 7.0, where Lint can run on multiple modules in parallel, cache incremental results, etc. If you try AGP 7.0 and things are still slow, feel free to file a bug at https://issuetracker.google.com/issues/new?component=192708&template=840533 and link it here. You can follow the instructions [here](https://github.com/google/android-lint-performance-probe/issues/5#issuecomment-610993161) to collect a YourKit CPU sampling snapshot.
